File: sdk/src/main/java/in/lazygod/sdk/TokenManager.java
package in.lazygod.sdk;

import com.fasterxml.jackson.databind.ObjectMapper;
import in.lazygod.sdk.dto.AuthRequest;
import in.lazygod.sdk.dto.AuthResponse;
import in.lazygod.sdk.dto.RefreshTokenRequest;
import in.lazygod.sdk.dto.RefreshTokenResponse;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Instant;

public class TokenManager {
    private final String baseUrl;
    private final String username;
    private final String password;
    private final HttpClient client = HttpClient.newHttpClient();
    private final ObjectMapper mapper = new ObjectMapper();

    private String accessToken;
    private String refreshToken;
    private Instant expiresAt;

    public TokenManager(String baseUrl, String username, String password) {
        this.baseUrl = baseUrl;
        this.username = username;
        this.password = password;
    }

    public synchronized String getAccessToken() throws IOException, InterruptedException {
        if (accessToken == null || Instant.now().isAfter(expiresAt)) {
            if (refreshToken != null) {
                refresh();
            } else {
                login();
            }
        }
        return accessToken;
    }

    private void login() throws IOException, InterruptedException {
        AuthRequest req = new AuthRequest(username, password);
        String json = mapper.writeValueAsString(req);
        HttpRequest http = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + "/auth/login"))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(json))
                .build();
        HttpResponse<String> resp = client.send(http, HttpResponse.BodyHandlers.ofString());
        if (resp.statusCode() >= 200 && resp.statusCode() < 300) {
            AuthResponse tokens = mapper.readValue(resp.body(), AuthResponse.class);
            this.accessToken = tokens.getAccessToken();
            this.refreshToken = tokens.getRefreshToken();
            // crude expiration assumption: 14 minutes
            this.expiresAt = Instant.now().plusSeconds(14 * 60);
        } else {
            throw new RuntimeException("Auth failed: " + resp.statusCode());
        }
    }

    private void refresh() throws IOException, InterruptedException {
        RefreshTokenRequest req = new RefreshTokenRequest(refreshToken);
        String json = mapper.writeValueAsString(req);
        HttpRequest http = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + "/auth/refresh"))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(json))
                .build();
        HttpResponse<String> resp = client.send(http, HttpResponse.BodyHandlers.ofString());
        if (resp.statusCode() >= 200 && resp.statusCode() < 300) {
            RefreshTokenResponse tokens = mapper.readValue(resp.body(), RefreshTokenResponse.class);
            this.accessToken = tokens.getAccessToken();
            this.refreshToken = tokens.getRefreshToken();
            this.expiresAt = Instant.now().plusSeconds(14 * 60);
        } else {
            login();
        }
    }
}

File: sdk/js/src/TokenManager.ts
import { AuthRequest, AuthResponse, RefreshTokenRequest, RefreshTokenResponse } from './types';

export default class TokenManager {
  private baseUrl: string;
  private username?: string;
  private password?: string;
  private accessToken?: string;
  private refreshToken?: string;
  private expiresAt = 0;

  constructor(baseUrl: string, username?: string, password?: string) {
    this.baseUrl = baseUrl;
    this.username = username;
    this.password = password;
  }

  async getAccessToken(): Promise<string> {
    if (!this.accessToken || Date.now() >= this.expiresAt) {
      if (this.refreshToken) {
        await this.refresh();
      } else {
        await this.login();
      }
    }
    return this.accessToken as string;
  }

  private async login(): Promise<void> {
    if (!this.username || !this.password) {
      throw new Error('Username/password not configured');
    }
    const req: AuthRequest = { username: this.username, password: this.password };
    const resp = await fetch(`${this.baseUrl}/auth/login`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(req)
    });
    if (!resp.ok) throw new Error(`Auth failed: ${resp.status}`);
    const data: AuthResponse = await resp.json();
    this.setTokens(data.accessToken, (data as any).refreshToken || data.refreshTooken);
  }

  private async refresh(): Promise<void> {
    const req: RefreshTokenRequest = { refreshToken: this.refreshToken as string };
    const resp = await fetch(`${this.baseUrl}/auth/refresh`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(req)
    });
    if (!resp.ok) {
      await this.login();
      return;
    }
    const data: RefreshTokenResponse = await resp.json();
    this.setTokens(data.accessToken, data.refreshToken);
  }

  setTokens(access: string, refresh: string) {
    this.accessToken = access;
    this.refreshToken = refresh;
    this.expiresAt = Date.now() + 14 * 60 * 1000;
  }
}
